src/riak_kv_w1c_worker.erl

```diff
@@ -354,3 +349,82 @@ erase_request_record(ReqId, #state{entries=Entries} = State) ->
         OldValue ->
             {OldValue, State#state{entries=?DICT_TYPE:erase(ReqId, Entries)}}
     end.
+
+%%%% Functions to handle asynchronous puts
+
+%% Invoked by put/2 to turn the async request into a synchronous call
+synchronize_put({ok, {ReqId, Worker}}, Options) ->
+    Timeout = case proplists:get_value(timeout, Options, ?DEFAULT_TIMEOUT) of
+                  N when is_integer(N) andalso N > 0 ->
+                      N;
+                  _ ->
+                      ?DEFAULT_TIMEOUT
+              end,
+    %% this is not direct gen_server:call because the process that
+    %% replies is not the one that received the cast
+    receive
+        {'DOWN', ReqId, process, _Pid, _Reason} ->
+            {error, riak_kv_w1c_server_crashed};
+        {ReqId, Response} ->
+            erlang:demonitor(ReqId, [flush]),
+            Response
+    after Timeout ->
+            gen_server:cast(Worker, {cancel, ReqId}),
+            receive
+                {'DOWN', ReqId, process, _Pid, _Reason} ->
+                    {error, riak_kv_w1c_server_crashed};
+                {ReqId, Response} ->
+                    erlang:demonitor(ReqId, [flush]),
+                    Response
+            end
+    end;
+synchronize_put(Error, _Options) ->
+    Error.
+
+%% Invoked by async_put_reply/2 to wait for all responses
+async_put_reply_loop([], Responses, _Timeout) ->
+    Responses;
+async_put_reply_loop(ReqIds, Responses, Timeout) ->
+    receive
+        {'DOWN', ReqId, process, _Pid, _Reason} when is_reference(ReqId) ->
```

Comment:
Oh, bloody ‘ell, you’re right. This was easier when we were expecting one specific ID.

Good catch, thanks.

-John

> On Nov 4, 2015, at 3:15 PM, Micah <notifications@github.com> wrote:
> 
> In src/riak_kv_w1c_worker.erl <https://github.com/basho/riak_kv/pull/1254#discussion_r43933086>:
> 
> > +                {'DOWN', ReqId, process, _Pid, _Reason} ->
> > +                    {error, riak_kv_w1c_server_crashed};
> > +                {ReqId, Response} ->
> > +                    erlang:demonitor(ReqId, [flush]),
> > +                    Response
> > +            end
> > +    end;
> > +synchronize_put(Error, _Options) ->
> > +    Error.
> > +
> > +%% Invoked by async_put_reply/2 to wait for all responses
> > +async_put_reply_loop([], Responses, _Timeout) ->
> > +    Responses;
> > +async_put_reply_loop(ReqIds, Responses, Timeout) ->
> > +    receive
> > +        {'DOWN', ReqId, process, _Pid, _Reason} when is_reference(ReqId) ->
> Except it doesn't seem to here (It does in the async_put function, so no worries there). We have a list of ReqIds in the function head, the receive clause just checks that ReqId is a reference (which all monitor's are), and blindly deletes from the ReqIds list, which will work even if the key doesn't exist. Unless I've completely mis-read the code...
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/basho/riak_kv/pull/1254/files#r43933086>.
>